### PR TITLE
1714: CSR detection only happens on the title bug of a PR

### PR DESCRIPTION
**Pull request by zhaosongzs**
Currently, our bots only detect CSR issues that are directly linked to the main issue. 

With this patch, our bots will be able to detect CSR issues that are linked to all of the issues that this PR solves. Additionally, all detected CSR issues will affect the integration blockers.

The `/csr needed` command will allow you to create a CSR issue that is linked to any of the issues solved by this PR. The `/csr unneeded` command will be blocked if any of the issues associated with this PR have an active CSR issue.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1714](https://bugs.openjdk.org/browse/SKARA-1714): CSR detection only happens on the title bug of a PR

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara pull/1443/head:pull/1443` \
`$ git checkout pull/1443`

Update a local copy of the PR: \
`$ git checkout pull/1443` \
`$ git pull https://git.openjdk.org/skara pull/1443/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1443`

View PR using the GUI difftool: \
`$ git pr show -t 1443`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1443.diff">https://git.openjdk.org/skara/pull/1443.diff</a>

</details>

### CODE REVIEW

File: bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java [hunk 1]
```diff
@@ -144,81 +146,103 @@ public Collection<WorkItem> run(Path scratchPath) {
...
-                log.info("CSR issue resolution name is null for " + describe(pr) + ", not removing the CSR label");
+            var csr = csrOptional.get();
+
+            log.info("Found CSR " + csr.id() + " for issue " + jbsIssueOpt.get().id() + " for " + describe(pr));
+            if (!hasCsrIssueAndProgress(pr, csr) && !isWithdrawnCSR(csr)) {
+                // If the PR body doesn't have the CSR issue or doesn't have the CSR progress,
+                // this bot need to add the csr update marker so that the PR bot can update the message of the PR body.
+                log.info("The PR body doesn't have the CSR issue or progress, adding the csr update marker for this csr issue"
+                        + csr.id() + " for " + describe(pr));
+                addUpdateMarker(pr);
```
Comment (erikj79) [inline bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java:175]: `addUpdateMarker` isn't suitable to be called multiple times. Each call will modify the body again.

File: bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java [hunk 2]
```diff
@@ -144,81 +146,103 @@ public Collection<WorkItem> run(Path scratchPath) {
...
-        if (resolution == null || resolution.isNull()) {
-            if (!pr.labelNames().contains(CSR_LABEL)) {
-                log.info("CSR issue resolution is null for " + describe(pr) + ", adding the CSR label");
-                pr.addLabel(CSR_LABEL);
-            } else {
-                log.info("CSR issue resolution is null for " + describe(pr) + ", not removing the CSR label");
+            var csrOptional = Backports.findCsr(jbsIssueOpt.get(), versionOpt.get());
+            if (csrOptional.isEmpty()) {
+                log.info("No CSR found for issue " + jbsIssueOpt.get().id() + " for " + describe(pr));
+                continue;
```
Comment (erikj79) [inline bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java:165]: What happens if no issue has a CSR but the label was added using the CSR command? In that case we should not remove the label.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java [hunk 3]
```diff
@@ -54,9 +55,9 @@ private static void jbsReply(PullRequest pr, PrintWriter writer) {
                       "an issue in JBS, please use the `/issue` command in a comment in this pull request.");
     }
 
-    private static void linkReply(PullRequest pr, Issue issue, PrintWriter writer) {
-        writer.println("@" + pr.author().username() + " please create a [CSR](https://wiki.openjdk.org/display/csr/Main) request for issue " +
-                "[" + issue.id() + "](" + issue.webUrl() + ") with the correct fix version. " +
+    private static void csrCreateReply(PullRequest pr,  PrintWriter writer) {
+        writer.println("@" + pr.author().username() + " please create a [CSR](https://wiki.openjdk.org/display/csr/Main) " +
+                "request for any issue associated with this pr with the correct fix version " +
                 "This pull request cannot be integrated until the CSR request is approved.");
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java:61]: This message doesn't really make sense. If there is more than one issue associated with the pr, then I don't think there is any point printing this at all. We can't know which issue that needs a CSR, so the person issuing the command has to be responsible for clarifying that. The main point of this message is to link to the issue.

So, if there is only one issue (which will be the case in 99% of cases involving CSRs), print the old message, otherwise don't call this method, or have it do nothing.

**Review by erikj79 — COMMENTED** [review]
I've made a first pass over this patch. It's a rather complicated change which introduces a lot of corner cases that need careful reasoning and testing. I found some issues that I've pointed out, but there could definitely be more. We also know that the current CSR implementation has issues, which may get amplified by this change. I'm still not sure if we should risk going ahead with this before we figure out a better solution to the whole CSR functionality.

File: bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java (hunk 1 above)
Comment (zhaosongzs) [reply bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java:175]: Right, I will change it, thx!

File: bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java (hunk 2 above)
Comment (zhaosongzs) [reply bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java:165]: I overlooked this case. Will fix it.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java [hunk 4]
```diff
@@ -96,109 +97,129 @@ public void handle(PullRequestBot bot, PullRequest pr, CensusInstance censusInst
...
+            } else if (csr.state() == Issue.State.CLOSED && resolutionName.equals("Withdrawn")) {
+                continue;
+            } else {
+                reply.println("this pull request will not be integrated until the [CSR](https://wiki.openjdk.org/display/csr/Main) " +
+                        "request " + "[" + csr.id() + "](" + csr.webUrl() + ")" + " for issue " +
+                        "[" + jbsIssue.id() + "](" + jbsIssue.webUrl() + ") has been approved.");
+                reply.println(CSR_NEEDED_MARKER);
+                pr.addLabel(CSR_LABEL);
+            }
+            return;
```
[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java:217]: Yes, I think this behavior is right because if we could reach the return statement, it means we already found a non-withdrawn CSR.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 5]
```diff
@@ -229,27 +243,53 @@ private List<String> botSpecificChecks(boolean iscleanBackport) {
...
+                .filter(issue -> issue.properties().containsKey("issuetype"))
+                .filter(issue -> issue.properties().get("issuetype").asString().equals("CSR"))
+                .filter(issue -> !isWithdrawnCSR(issue))
+                .toList();
+        if (csrIssues.isEmpty() && pr.labelNames().contains("csr")) {
+            ret.put("Change requires a CSR request (need to be created) to be approved", false);
+        }
+        for (var csrIssue : csrIssues) {
+            if (!csrIssue.isClosed()) {
+                ret.put("Change requires a CSR request (" + csrIssue.id() + ") to be approved", false);
```
[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:272]: I thought about this too. But then I realized the links can be found under the 'Issues' part in PR body. Do you think we still need the links here? If so, I will add the links.

File: bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java [hunk 6]
```diff
@@ -126,15 +129,14 @@ private void addUpdateMarker(PullRequest pr) {
     @Override
     public Collection<WorkItem> run(Path scratchPath) {
         var pr = repository.pullRequest(prId);
+        // All the issues this pr solves
+        var mainIssue = org.openjdk.skara.vcs.openjdk.Issue.fromStringRelaxed(pr.title());
+        var issues = new ArrayList<org.openjdk.skara.vcs.openjdk.Issue>();
+        mainIssue.ifPresent(issues::add);
+        issues.addAll(SolvesTracker.currentSolved(pr.repository().forge().currentUser(), pr.comments()));
```
[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/csr/src/main/java/org/openjdk/skara/bots/csr/PullRequestWorkItem.java:136]: Thanks, fixed it.

**Comment by zhaosongzs** [comment]
> I've made a first pass over this patch. It's a rather complicated change which introduces a lot of corner cases that need careful reasoning and testing. I found some issues that I've pointed out, but there could definitely be more. We also know that the current CSR implementation has issues, which may get amplified by this change. I'm still not sure if we should risk going ahead with this before we figure out a better solution to the whole CSR functionality.

I fixed some of the problems you mentioned.  

Before trying to fix SKARA-1714, I didn't realize it's so complicated. So I can totally understand your concerns. There must be some corner cases we haven't covered. I think I should do more tests and also think of how to move the functions in CSR bots to PR bots later.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java:218]: But what if there are more?

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java]: I think it would help.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java:218]: If we found one approved, we will reply some messages to user and return. What do you think should we do if there is another that is withdrawn? I don't think we need to take any action.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java]: Ok, I will add the links

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (erikj79) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java:218]: Say there are two CSRs present already, one approved and one open. If the approved one is found first, then we get the needed marker but no label. If the open one is found first, then we also get the label. This is inconsistent behavior.

What I think should happen is.

If there are already CSRs present, then
1. if all are approved, just add the needed marker
2. If at least one is open, add both needed marker and label
3. If all are withdrawn, add both needed marker and label

To achieve this you will have to always iterate over all of them and set booleans for each condition, then decide the correct action based on the complete state.

[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/pr/src/main/java/org/openjdk/skara/bots/pr/CSRCommand.java:218]: Yes, you are right. I will fix it.

**Comment by zhaosongzs** [comment]
> I think this is ok now, but would be good if you could try it in staging on the playground repo with bugs-stage and go through some scenarios.

Thanks for the review! Sure, I would do some tests in staging later.
